### PR TITLE
feat: reduce search providers per config value "unified_search_providers_allowed"

### PR DESCRIPTION
**Pull request by printminion-co**
## Summary
This feature should help to restrict not desired search providers by allowing only desired search providers.

### Before
One can see  the `/ocs/v2.php/search/providers` deliver various search providers (you may see other providers on your nextcloud instance depending on installed/disabled apps).
Selection_20241023-002

### After feature activation
One can see we allowed only the `files` and `settings` search providers (since we do not want to allow any other search providers).

Selection_20241023-003

CLI Example: restrict search providers to `files` and `settings`

```shell
./occ config:app:set --value '["files","settings"]' --type array core unified_search.providers_allowed
```

* Resolves: # 

## TODO

- [x] Disable search result on "disabled" providers e.g. `/ocs/v2.php/search/providers/settings/search?term=f&from=%2Fapps%2Ffiles%2Ffiles`

## Checklist

- Code is [properly formatted](https://docs.nextcloud.com/server/latest/developer_manual/digging_deeper/continuous_integration.html#linting)
- [Sign-off message](https://github.com/src-d/guide/blob/master/developer-community/fix-DCO.md) is added to all commits
- [ ] Tests ([unit](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#unit-tests), [integration](https://docs.nextcloud.com/server/latest/developer_manual/app_development/tutorial.html#integration-tests), api and/or acceptance) are included
- [x] Screenshots before/after for front-end changes
- [ ] Documentation ([manuals](https://github.com/nextcloud/documentation/) or wiki) has been updated or is not required
- [ ] [Backports requested](https://github.com/nextcloud/backportbot/#usage) where applicable (ex: critical bugfixes)

### CODE REVIEW

File: lib/private/Search/SearchComposer.php [hunk 1]
```diff
@@ -197,7 +202,24 @@ function (array $providerData) use ($route, $routeParameters) {
...
-		return $providers;
+		return $this->reduceProviders($providers);
+	}
+
+	/**
+	 * reduce providers based on 'unified_search_providers_allowed' config array
+	 * @param array $providers
+	 * @return array
+	 */
+	private function reduceProviders(array $providers): array {
```
Comment (artonge) [inline lib/private/Search/SearchComposer.php:213]:
```suggestion
	private function filterProviders(array $providers): array {
```

**Review by artonge — COMMENTED** [review]
Is the endpoint also not responding when searching in disabled providers?

**Review by susnux — COMMENTED** [review]
I am not sure, as far as I know we do not have documentation about available search providers. So this config would be blind-guessing.
Maybe it is safer to make it an app config first until we have a occ command to list search providers?

**Comment by sorbaugh** [comment]
> would be blind-guessing. Maybe it is safer to make it an app config first until we have a occ command to list search providers?

@printminion-co would this be ok for you?

**Comment by printminion-co** [comment]
> I am not sure, as far as I know we do not have documentation about available search providers. So this config would be blind-guessing. Maybe it is safer to make it an app config first until we have a occ command to list search providers?

I've added more description and screenshots to better describe the feature.

cc// @artonge @susnux @sorbaugh

**Comment by susnux** [comment]
> We know which providers we want keep enabled.
> We do not need a listing.

For you this might be true, but my fear is that this will confuse other administrators of Nextcloud systems, because of the missing listing / documentation.

>  There is no blind-guessing since we allow and not disable the search providers.

Well if you know that API endpoint to retrive the available providers yes. What I meant is:
There is no list of available providers in the documentation and no OCC command to query it,
so this would be a feature mostly hidden and undocumented.

That is why I asked if instead of system config also an app config would work? Instead of writing it to the config.php just have it as an `core` app setting, we could later built an OCC command around it (listing + dis-/enable).

From my side no blocker, but just wanted to bring this up as this might lead to confusion in the future.

File: config/config.sample.php [hunk 2]
```diff
@@ -2534,6 +2534,15 @@
...
 'unified_search.enabled' => false,
 
+/**
+ * List of search providers that are allowed to be used in the unified search
+ *
+ * Defaults to ``[]`` (all providers allowed)
+ *
+ * For example: Set to ['files', 'settings'] to only allow file and settings search.
+ */
+'unified_search_providers_allowed' => [],
```
Comment (susnux) [inline config/config.sample.php:2544]: We should keep the same format for all related options:

```suggestion
'unified_search.providers_allowed' => [],
```

**Comment by printminion-co** [comment]
> Instead of writing it to the config.php just have it as an `core` app setting, we could later built an OCC command around it (listing + dis-/enable).

I will check this

**Comment by printminion-co** [comment]
> Instead of writing it to the config.php just have it as an `core` app setting, we could later built an OCC command around it (listing + dis-/enable).

do you mean here to use 
```php
$allowedProviders = $this->appConfig->getValueArray('core', 'unified_search.providers_allowed');
```
which is set via
```shell
./occ config:app:set --value '["files","settings"]' --type array core unified_search.providers_allowed
```
@susnux

File: lib/private/Search/SearchComposer.php (hunk 1 above)
Comment (printminion-co) [reply lib/private/Search/SearchComposer.php]: changed

**Comment by artonge** [comment]
Looks good to me. Could you answer my previous question? https://github.com/nextcloud/server/pull/48841#pullrequestreview-2385467360

**Comment by printminion-co** [comment]
> Looks good to me. Could you answer my previous question? [#48841 (review)](https://github.com/nextcloud/server/pull/48841#pullrequestreview-2385467360)

Not sure what "endpoint" mean. 
Do you mean the API call of `/ocs/v2.php/search/providers` will deliver reduced providers? If so yes, one can observe it in the screenshots in PR description.

**Comment by printminion-co** [comment]
> > Not sure what "endpoint" mean. Do you mean the API call of `/ocs/v2.php/search/providers` will deliver reduced providers? If so yes, one can observe it in the screenshots in PR description.
> 
> No, I meant if you make requests to the search controllers, I suspect that disabled providers will still answer results.

You are right... ..the GET call to "disabled" controller `/ocs/v2.php/search/providers/settings/search?term=f&from=%2Fapps%2Ffiles%2Ffiles` still delivers a result. 

I should also disable it

**Comment by printminion-co** [comment]
> > Not sure what "endpoint" mean. Do you mean the API call of `/ocs/v2.php/search/providers` will deliver reduced providers? If so yes, one can observe it in the screenshots in PR description.
> 
> No, I meant if you make requests to the search controllers, I suspect that disabled providers will still answer results.

I moved the logic into the `loadProviders` Method

Now API call to disabled provider delivers same response as request to not existing search provider e.g. `ocs/v2.php/search/providers/foo/search?term=f&from=%2Fapps%2Ffiles%2Ffiles`

```xml
<?xml version="1.0"?>
<ocs>
 <meta>
  <status>failure</status>
  <statuscode>996</statuscode>
  <message>Internal Server Error
</message>
 </meta>
 <data/>
</ocs>
```

**Review by come-nc — APPROVED** [review]

**Review by nfebe — APPROVED** [review]
The admin documentation should mention somewhere that this is possible and how! Nice to have is probably a new follow issue covers the complete implementation of the occ command with list and dis/enable features.

**Review by artonge — APPROVED** [review]

**Comment by artonge** [comment]
Looks good @printminion-co. Can you rebase?

**Pull request merged**